### ISSUE TITLE
Evaluate automated dependency upgrade products

### ISSUE THREAD

**Issue by rianfowler** [department-of-veterans-affairs/va.gov-team#9537]
- We currently rely on dependabot for automated dependency upgrades 
  - Doesn't provide a lot of feedback on dependencies in general 
  - Process / tooling is not documented. No policy around dependency upgrades is documented 
- Evaluate other automated dependency upgrade products 
  - [depfu](https://depfu.com/?utm_source=thenewstack&utm_medium=website) for example has a dependency upgrade scheduler 
- create a policy for dependency management 
- determine how to monitor / alert for upgrades if not available from automated tools

**Comment by jhonnyoddball** [comment]
I'll be looking at these dependency bots for further evaluation:

- [Snyk](https://snyk.io/)
- [Greenkeeper](https://greenkeeper.io/)
- [Renovate](https://renovate.whitesourcesoftware.com/)
- [Depfu](https://depfu.com/)

**Comment by jhonnyoddball** [comment]
## Snyk

Snyk automatically detects open source vulnerabilities. We can also look into this to improve the security vulnerabilities.

It's not design to handle `npm` automatic upgrades.

**Comment by jhonnyoddball** [comment]
# Depfu

`Depfu` is a platform that will help to keep the repo(s) dependencies up-to-date.

By default `Depfu` creates one PR per new version, but it can also get weekly or monthly PRs updating several dependencies at once.

`Depfu` figures out the minimum number of updates needed to keep the repo reasonably up-to-date. `Depfu` creates new versions depending on the library's past release frequency instead of opening a PR right away. This reduces the number of PRs per week compared to creating PRs for every new version.

`Depfu` provides a dashboard to manage repos and its independent settings

### Dashboard
 
![Defu Dashboard](https://user-images.githubusercontent.com/55560129/83437826-6b01d100-a40e-11ea-9801-c716d0414a47.png)

### Settings

![Depfu settings](https://user-images.githubusercontent.com/55560129/83437706-37bf4200-a40e-11ea-9253-72dc8cab6a17.png)

### Add a Repo

We can add all repos or add selected repos

![Depfu add repo](https://user-images.githubusercontent.com/55560129/83438120-de0b4780-a40e-11ea-81d2-f83e84a599da.png)

<HR>

# Repos

### Repo Dashboard
 
![Depfu repo dashboard](https://user-images.githubusercontent.com/55560129/83436615-2c6b1700-a40c-11ea-8eba-290929b0d317.png)

### Repo Settings

These settings can be adjusted per repo

![Depfu repo settings](https://user-images.githubusercontent.com/55560129/83438799-ff206800-a40f-11ea-8c5b-39516e1dd525.png)

Unfortunately, due to limitations with GitHub, we won't be able to assign the new PRs to a team

![Depfu Assignee policy](https://user-images.githubusercontent.com/55560129/83438951-38f16e80-a410-11ea-992d-ff42221d0622.png)

### Automatic Updates

Once the repo is added to `Depfu`, it will start running and checking for possible dependency updates. It will start with the minor updates.

![Depfu PRs created](https://user-images.githubusercontent.com/55560129/83438977-43136d00-a410-11ea-8cb4-45b0e1b71652.png)

![Depfu PR details](https://user-images.githubusercontent.com/55560129/83438991-4870b780-a410-11ea-8a7d-6fdf6b86e789.png)

### Manual Updates

If a dependency update is needed right away, the PR can be created manually using `Depfu`.

![Depfu create PR](https://user-images.githubusercontent.com/55560129/83438287-29bdf100-a40f-11ea-8adb-e7ae7ec0cc3b.png)

![Depfu Manual update](https://user-images.githubusercontent.com/55560129/83438297-2f1b3b80-a40f-11ea-816b-bc0e99ffd627.png)

![Depfu manual update PR](https://user-images.githubusercontent.com/55560129/83438307-33dfef80-a40f-11ea-93f8-46ca75390302.png)

**Comment by jhonnyoddball** [comment]
# Renovate

Renovate is a GitHub application that monitors dependency updates for a given repository and opens Pull Requests when new versions are available.

Renovate uses a configuration file, `renovate.json`, that lives at the root of each repository to handle how Renovate manages the dependencies for that project. These files behave similar to an `.eslintrc` file, providing multiple presets and reasonable defaults for various use cases using the extends property.

#### `renovate.json` example

```
{
  "extends": ["config:base"],
  "baseBranches": ["develop"],
  "packageRules": [
    {
      "depTypeList": ["devDependencies"],
      "updateTypes": ["minor"],
      "automerge": true
    }
  ],
  "prHourlyLimit": 5,
  "prConcurrentLimit": 10,
  "rangeStrategy": "bump"
}
```

A key differentiating factor that Renovate provides is the ability for fine-grained control of automerging dependencies based on rules set in the config. Automerging is disabled by default, but can be a useful tool for keeping lower risk dependencies updated with reduced friction.

Another benefit of Renovate, is the grouping of monorepo updates in a single PR. Creating PRs for each dependency can cause problems with the proliferation of monorepos such as Babel and React. In most cases, these dependencies need to be updated at the same time or they can break the application.

A downside so far is the need to request access to a `department-of-veterans-affairs` owner to be able to create the initial PR setup.

![Screen Shot 2020-06-02 at 9 53 03 AM](https://user-images.githubusercontent.com/55560129/83555020-a7e7c980-a4db-11ea-9767-509feb84a6f8.png) 

### Resources

- [Documentation](https://docs.renovatebot.com/)
- [Configuration options](https://docs.renovatebot.com/configuration-options/)

### Dashboard

The dashboard is very simple, there really isn't anything that can be done here. Everything is done in the `renovate.json` file.

![Renovate Dashboard](https://user-images.githubusercontent.com/55560129/83556206-7ff96580-a4dd-11ea-9bc6-df151f5ce8ba.png)

![Renovate job details](https://user-images.githubusercontent.com/55560129/83556216-84258300-a4dd-11ea-82c1-9d8a54cdc838.png)

### Automatic Updates

Once Renovate is added to the repo, it will create an initial PR. Once the PR is merged, Renovate will start creating all the necessary PRs based on the configurations.

![Renovate initial PR ](https://user-images.githubusercontent.com/55560129/83556675-55f47300-a4de-11ea-85c1-c8ef2869a63e.png)

![Renovate initial PR details](https://user-images.githubusercontent.com/55560129/83556695-5c82ea80-a4de-11ea-96ff-603417d99735.png)

**PRs after initial PR is merged**

![Renovate update PR](https://user-images.githubusercontent.com/55560129/83558289-cd2b0680-a4e0-11ea-8150-e5f47de4259e.png)

![Renovate update PR details](https://user-images.githubusercontent.com/55560129/83558304-d1572400-a4e0-11ea-8d10-f9cc34a5e722.png)

**If an update PR is closed and not merged. Renovate will recreate a new PR**

![Renovate PRs](https://user-images.githubusercontent.com/55560129/83558655-56423d80-a4e1-11ea-992a-9085f479619c.png)

### Configurations

Renovate provides an extensive amount of [configuration options](https://docs.renovatebot.com/configuration-options/).

These are some options worth testing:

- [automerge](https://docs.renovatebot.com/configuration-options/#automerge)
- [labels](https://docs.renovatebot.com/configuration-options/#labels)
- [lockFileMaintenance](https://docs.renovatebot.com/configuration-options/#lockfilemaintenance)
- [packageRules](https://docs.renovatebot.com/configuration-options/#packagerules)
- [prConcurrentLimit](https://docs.renovatebot.com/configuration-options/#prconcurrentlimit)
- [prHourlyLimit](https://docs.renovatebot.com/configuration-options/#prhourlylimit)
- [rangeStrategy](https://docs.renovatebot.com/configuration-options/#rangestrategy)
- [reviewersFromCodeOwners](https://docs.renovatebot.com/configuration-options/#reviewersfromcodeowners)
- [schedule](https://docs.renovatebot.com/configuration-options/#schedule)

**Comment by jhonnyoddball** [comment]
# Dependabot

`Dependabot` was initially installed in `vets-website` but it wasn't doing anything. There was no action by `Dependabot`, therefore, we started looking at alternatives.

![Screen Shot 2020-06-04 at 4 39 31 PM](https://user-images.githubusercontent.com/55560129/83808043-0cdd2400-a682-11ea-9a07-2455094fb4db.png)

I got access to the `dependabot` dashboard and noticed that it was not longer there. Someone removed it. I have added it back and now I'm testing it.

`dependabot` has good settings and also [it's been integrated into GitHub](https://dependabot.com/blog/hello-github/). Now, that is integrated, we will have no problem requesting access to add the app.

`dependabot` uses a configuration file, `dependabot.yml`, that lives at the root of each repository to handle how Renovate manages the dependencies for that project.

#### `dependabot.yml` example

```
version: 2
updates:
  - package-ecosystem: "npm"
    directory: "/"
    schedule:
      interval: "daily"
      time: "09:00"
      timezone: "Asia/Tokyo"
```

### Supported Languages

`Dependabot` monitors security advisories for Ruby, JavaScript, PHP, Java, . NET, Python, Elixir and Rust.

### Resources

- [Documentation](https://help.github.com/en/github/administering-a-repository/keeping-your-dependencies-updated-automatically)

### Dashboard

![Dependabot dashboard](https://user-images.githubusercontent.com/55560129/83884447-e5826780-a712-11ea-95da-00d1cd598d11.png)

### Settings

![Dependabot Settings](https://user-images.githubusercontent.com/55560129/83884710-4ca01c00-a713-11ea-8bb0-eb578077982c.png)

### Add a Repo

We can any repo within the organization

![Dependabot add repo](https://user-images.githubusercontent.com/55560129/83884819-76f1d980-a713-11ea-9008-0eba5b18e6c7.png)

<HR>

# Repos

### Repo Settings

These settings can be adjusted per repo

![Dependabot repo settings](https://user-images.githubusercontent.com/55560129/83884886-8f61f400-a713-11ea-8ca5-95450bb82fdc.png)

### Automatic Updates

Once the repo is added to `Dependabot`, it will start running and checking for possible dependency updates. 

![Dependabot PRs](https://user-images.githubusercontent.com/55560129/83884991-bb7d7500-a713-11ea-89e1-547ae9a248d6.png)

### Manual Updates

You can't run manual updates per dependency, however, you can run the job with the `bump now` button.

### Ignore PR

If a PR is generated by `Dependabot` and then closed it will be ignored until a new version is available. Also, an email is sent with that information

![Dependabot closed PR](https://user-images.githubusercontent.com/55560129/83886495-224f5e00-a715-11ea-98e7-7e10410721ad.png)

**Comment by jhonnyoddball** [comment]
I would recommend testing `Dependabot` inside the `vets-website` and see how good it is on keeping up-to-date all the dependencies. Now, that `Dependabot` is being integrated inside GitHub, there is no need to add any extra application.

Also, if we are to implement the micro front-ends to separate all the apps. If we use lerna, we will have no problem [running `Dependabot` on a monorepo](https://dependabot.com/blog/lerna-support/).

**Comment by jhonnyoddball** [comment]
cc @gunsch  @drorva

**Comment by rianfowler** [comment]
@gunsch @drorva I'm going to close this ticket. We're going to stick with using dependabot for managing our dependency upgrades. We'll continue to make this part of the[ support rotation responsibility](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/teams/vsp/teams/tools/frontend/support-and-triage.md) on FE.
